### ISSUE TITLE
Protocol error in kinesis putRecords in aws sdk > 3.81.0

### ISSUE THREAD

**Issue by cesarfd** [aws/aws-sdk-js-v3#3780]
### Describe the bug

This is a reopening of #3602 as the bug is definitely not fixed as of v3.121.0

### Expected Behavior

putRecords() should work correctly.

### Current Behavior

Connection fails, error msg gives `"Protocol error"`.

### Reproduction Steps

Please refer to https://github.com/aws/aws-sdk-js-v3/issues/3602#issuecomment-1134973121

### SDK version used

3.82.0 or later

### Environment details (OS name and version, etc.)

Please refer to https://github.com/aws/aws-sdk-js-v3/issues/3602#issuecomment-1134973121

**Comment by gugu** [comment]
I confirm that it is not fixed.
Here is what I have in Sentry (not much)

<img width="645" alt="image" src="https://user-images.githubusercontent.com/75169/178790326-99b628f8-2c84-4f59-9c31-04f314a63b76.png">

**Comment by gugu** [comment]
@cesarfd let's try to debug. Do you use some other other AWS service, which does not support http2, for example DynamoDB? This error happens if someone uses http2 protocol for the endpoint, which does not support http2. And my guess aws sdk accidentally upgrades protocol version of the other service, and it fails

**Comment by cesarfd** [comment]
I can confirm it fails with and without using a custom VPC endpoint for Kinesis Streams (I tried it just in case it made a difference).

One thing that I don't understand is why is this not failing to everyone? AFAIK there should be nothing special that has to be configured in the client side or in the AWS console to use http/2 with any of the services, right?

**Comment by gugu** [comment]
Here is my workaround:
```
    const usKinesis = new KinesisClient({ region: "us-east-1", requestHandler: new NodeHttpHandler() });
```

**Comment by gugu** [comment]
My guess it is because we are the only people who use both http2-enabled AWS services and http1-only AWS services together

**Comment by cesarfd** [comment]
> Here is my workaround:
> 
> ```
>     const usKinesis = new KinesisClient({ region: "us-east-1", requestHandler: new NodeHttpHandler() });
> ```

I can confirm the workaround works. At least we can force it to use classic http 1.1

**Comment by RanVaknin** [comment]
@gugu 
Thanks for posting the workaround.

I added a p2 tag to it. Since it has a workaround other more pressing issues will take precedence, but this will be added to the team's backlog.

Thanks again,
Ran

**Comment by 300LiterPropofol** [comment]
> NodeHttpHandler

Hi @gugu, thank you for the post. I have one dumb question, where does the `new NodeHttpHandler()` come from? What `require()` do I need to import to use this handler or is it built-in in Nodejs? Thank you!

**Comment by dorrogeray** [comment]
@300LiterPropofol 
```ts
import {NodeHttpHandler} from '@aws-sdk/node-http-handler'
```

**Comment by jomaora** [comment]
I have the same error on v3.405.0, with NodeJS v18.17.0

```js
const document = // DATA TO SEND

const kinesis = new KinesisClient({
  endpoint: 'http://localhost:4567',
  credentials: {
    secretAccessKey: 'local',
    accessKeyId: 'local'
  },
  region: 'eu-west-1'
});

const command = new PutRecordCommand({
        StreamName: 'streamName',
        Data: Buffer.from(JSON.stringify(document)),
        PartitionKey: document._id
});

const result = await kinesis.send(command);
```

```
  Error (NghttpError) {
    $metadata: {
      attempts: 1,
      totalRetryDelay: 0,
    },
    code: 'ERR_HTTP2_ERROR',
    errno: -505,
    message: 'Protocol error',
  }
```

**Comment by a8t** [comment]
^ Me too, sdk `3.405.0` and node 20.6.0

**Comment by godu** [comment]
> I have the same error on v3.405.0, with NodeJS v18.17.0
> 
> ```js
> const document = // DATA TO SEND
> 
> const kinesis = new KinesisClient({
>   endpoint: 'http://localhost:4567',
>   credentials: {
>     secretAccessKey: 'local',
>     accessKeyId: 'local'
>   },
>   region: 'eu-west-1'
> });
> 
> const command = new PutRecordCommand({
>         StreamName: 'streamName',
>         Data: Buffer.from(JSON.stringify(document)),
>         PartitionKey: document._id
> });
> 
> const result = await kinesis.send(command);
> ```
> 
> ```
>   Error (NghttpError) {
>     $metadata: {
>       attempts: 1,
>       totalRetryDelay: 0,
>     },
>     code: 'ERR_HTTP2_ERROR',
>     errno: -505,
>     message: 'Protocol error',
>   }
> ```

We found a workaround. It was because we used [mhart/kinesalite](https://github.com/mhart/kinesalite) to simulate Kinesis locally. Unfortunately, [kinesalite doesn't support HTTP2 connect](https://github.com/mhart/kinesalite/issues/118#issuecomment-1697203599). We had to configure kinesis client to use HTTP1.1 instead.

```typescript
import {NodeHttpHandler} from '@smithy/node-http-handler';
import {KinesisClient} from '@aws-sdk/client-kinesis';

const kinesis = new KinesisClient(
  KINESIS_ENDPOINT
    ? {
        endpoint: KINESIS_ENDPOINT,
        requestHandler: new NodeHttpHandler()
      }
    : {}
);
```

**Comment by a8t** [comment]
Sorry, i neglected to say that the same workaround worked for us:

```nodejs
import { KinesisClient } from '@aws-sdk/client-kinesis';
import { NodeHttpHandler } from '@smithy/node-http-handler';

const region = 'us-east-2';
const kinesisClient = new KinesisClient({
  region,
  requestHandler: new NodeHttpHandler(),
  maxAttempts: 5,
});

export default kinesisClient;
```

This workaround was mentioned earlier in the thread, but since then, AWS has moved some libraries under `@smithy`.

**Comment by mitzafon-monday** [comment]
Still happening on `3.418.0` and now `@smithy` types aren't compatible with `NodeHttpHandler` when used in `requestHandler`.

Any advice?

**Comment by RanVaknin** [comment]
Hi there,

I tried to reproduce this with the latest version, but I'm not hitting this error. Kinesis supports both http2 and http1.1, and the default httphandler for this service uses http2 (likely to support streaming). The workaround suggested here is setting the request handler to the default protocol which 1.1. This is a totally viable workaround as the service supports both.

If anyone here can come up with a way to force this error we can take another look, but as it stands it is not reproducible. 

Thanks,
Ran~
